### PR TITLE
feat(admin): add option to invite multiple members to Vo/Group at once

### PR DESCRIPTION
**Pull request by xflord**
* added new dialog that allows invitation of multiple members via CSV
* update OpenAPI

### CODE REVIEW

**Comment by xflord** [comment]
Merge after https://github.com/CESNET/perun/pull/3782

File: apps/admin-gui/src/assets/i18n/en.json [hunk 1]
```diff
@@ -2019,6 +2023,26 @@
       "REMOVE_SPONSOR_DISABLED": "You are not authorized to remove sponsor",
       "SUCCESS": "Sponsors successfully edited"
     },
+    "BULK_INVITE_MEMBERS": {
+      "TITLE": "Invite multiple members",
+      "INSERT_HERE": "Here insert correctly formatted text",
+      "DESCRIPTION_VO": "The persons you are inviting will receive an email with a link to organizations application form. You can set up an email template in: (Organization) Settings -> Notifications.",
```
Comment (Johaney-s) [inline apps/admin-gui/src/assets/i18n/en.json:2029]:
```suggestion
      "DESCRIPTION_VO": "The persons you are inviting will receive an email with a link to organization's application form. You can set up an email template in: (Organization) Settings -> Notifications.",
```
same with "group's" below

File: apps/admin-gui/src/assets/i18n/en.json [hunk 2]
```diff
@@ -2019,6 +2023,26 @@
...
+      "DESCRIPTION_GROUP": "The persons you are inviting will receive an email with a link to the groups application form. You can set up an email template in: (Group) Settings -> Notifications.",
+      "LANGUAGE": "Language",
+      "LANGUAGE_ENGLISH": "English",
+      "LANGUAGE_CZECH": "Czech",
+      "PLACEHOLDER": "'John;john@mail.cz' or 'john@mail.cz'",
```
Comment (Johaney-s) [inline apps/admin-gui/src/assets/i18n/en.json:2034]:
```suggestion
      "PLACEHOLDER": "'john@mail.cz;John' or 'john@mail.cz'",
```

File: apps/admin-gui/src/app/vos/pages/group-detail-page/group-members/group-members.component.html [hunk 3]
```diff
@@ -24,14 +24,32 @@ <h1 class="page-subtitle">{{'GROUP_DETAIL.MEMBERS.TITLE' | translate}}</h1>
...
-      {{'GROUP_DETAIL.MEMBERS.INVITE_MEMBER' | translate}}
+      {{'GROUP_DETAIL.MEMBERS.INVITE' | translate}}
     </button>
+    <mat-menu #menu="matMenu">
+      <button
+        (click)="onInviteMember()"
+        *ngIf="inviteAuth"
+        [disabled]="(group | groupMembersActionButtonDisabled)"
+        class="action-button"
+        mat-flat-button>
```
Comment (Johaney-s) [inline apps/admin-gui/src/app/vos/pages/group-detail-page/group-members/group-members.component.html:41]:
```suggestion
        mat-menu-item>
```

File: apps/admin-gui/src/app/vos/pages/group-detail-page/group-members/group-members.component.html [hunk 4]
```diff
@@ -24,14 +24,32 @@ <h1 class="page-subtitle">{{'GROUP_DETAIL.MEMBERS.TITLE' | translate}}</h1>
...
+        {{'GROUP_DETAIL.MEMBERS.INVITE_ONE' | translate}}
+      </button>
+      <button
+        *ngIf="inviteAuth"
+        (click)="onBulkInvite()"
+        [disabled]="(group | groupMembersActionButtonDisabled)"
+        class="action-button"
+        mat-flat-button>
```
Comment (Johaney-s) [inline apps/admin-gui/src/app/vos/pages/group-detail-page/group-members/group-members.component.html:49]:
```suggestion
        mat-menu-item>
```

File: apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.html [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+          required
+          matInput
+          placeholder="{{'DIALOGS.BULK_INVITE_MEMBERS.PLACEHOLDER' | translate}}"
+          rows="8">
+        </textarea>
+        <mat-error *ngIf="invitedMembers.hasError('required')">
+          {{'DIALOGS.BULK_INVITE_MEMBERS.NAMES_ERROR' | translate}}
+        </mat-error>
+        <mat-error *ngIf="invitedMembers.hasError('invalidFormat')">
+          {{'DIALOGS.BULK_INVITE_MEMBERS.ERROR_FORMAT'| translate}}:
```
Comment (Johaney-s) [inline apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.html:38]: nitpick- if there are too many values inputted, the message can be confusing. maybe different alert should be used in that case.

**Review by Johaney-s — CHANGES_REQUESTED** [review]
superb! :rocket: tested edge cases and it seems to be working well, only small suggestions from me.

File: apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+  status: string;
+}
+
+@Component({
+  selector: 'app-bulk-invite-members-dialog',
+  templateUrl: './bulk-invite-members-dialog.component.html',
+  styleUrls: ['./bulk-invite-members-dialog.component.scss'],
+})
+export class BulkInviteMembersDialogComponent implements OnInit {
+  invitedMembers = new UntypedFormControl('', [Validators.required, this.userInputValidator()]);
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.ts:31]: Please use `FormControl` instead and after remove redundant type casting.

File: apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+      if (resultData[key] !== 'OK') {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  ngOnInit(): void {
+    this.theme = this.data.theme;
+    this.languages = this.store.getProperty('supported_languages');
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.ts:60]: Don't duplicate the theme. Also assign the supported languages directly.

File: apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.ts [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+  onSubmit(): void {
+    this.loading = true;
+    const listOfMembers = (this.invitedMembers.value as string).split('\n');
+    const parsedMembers: string[] = [];
+    for (const line of listOfMembers) {
+      const parsedLine = this.parseMemberLine(line);
+      if (parsedLine !== 'format' && parsedLine !== 'email') {
+        if (parsedLine !== '') {
+          parsedMembers.push(parsedLine);
+        }
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.ts:76]: Do we even need this check here ? It should not be possible to click on submit when the format is invalid.

File: apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.ts [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+      error: () => {
+        this.loading = false;
+      },
+    });
+  }
+
+  downloadCsv(): void {
+    if (!this.resultData) {
+      throw new Error('Cannot generate pdf because there is no result');
+    }
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.ts:109]: Here again if such case can happen, disable the download button instead.

File: apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.html [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        {{'DIALOGS.BULK_INVITE_MEMBERS.DESCRIPTION_GROUP' | translate}}
+      </div>
+      <mat-form-field class="w-100 pt-2">
+        <mat-label>{{'DIALOGS.BULK_INVITE_MEMBERS.LANGUAGE' | translate}}</mat-label>
+        <mat-select [(value)]="currentLanguage">
+          <mat-option
+            *ngFor="let lang of languages"
+            value="{{lang}}"
+            >{{'SHARED_LIB.LANGUAGES.'+lang | uppercase | translate}}</mat-option
+          >
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.html:19]:
```suggestion
          <mat-option *ngFor="let lang of languages" value="{{lang}}">
            {{'SHARED_LIB.LANGUAGES.'+lang | uppercase | translate}}
          </mat-option>
```

File: apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.html [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+      {{'DIALOGS.BULK_INVITE_MEMBERS.SUCCESS' | translate}}
+    </perun-web-apps-alert>
+    <perun-web-apps-alert *ngIf="this.finishedWithErrors" [alert_type]="'warn'">
+      {{'DIALOGS.BULK_INVITE_MEMBERS.FINISHED_WITH_ERRORS' | translate}}
+    </perun-web-apps-alert>
+  </div>
+  <div mat-dialog-actions *ngIf="state === 'input'">
+    <button (click)="onCancel()" class="ml-auto" mat-flat-button>
+      {{'DIALOGS.BULK_INVITE_MEMBERS.CANCEL' | translate}}
+    </button>
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.html:63]: Now the cancel button is duplicated. We only need only div for the dialog actions and we can put the check on the buttons themself.

File: apps/admin-gui/src/assets/i18n/en.json (hunk 1 above)
Comment (bodnara) [inline apps/admin-gui/src/assets/i18n/en.json:2029]: Also
```suggestion
      "DESCRIPTION_VO": "The people you are inviting will receive an email with a link to organizations application form. You can set up an email template in: (Organization) Settings -> Notifications.",
```

**Review by bodnara — CHANGES_REQUESTED** [review]

File: apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.ts [hunk 12]
```diff
@@ -0,0 +1,142 @@
+import { Component, Inject } from '@angular/core';
+import { AbstractControl, FormControl, ValidatorFn, Validators } from '@angular/forms';
```
Comment (bodnara) [inline apps/admin-gui/src/app/shared/components/dialogs/bulk-invite-members-dialog/bulk-invite-members-dialog.component.ts:2]:
```suggestion
import { FormControl, ValidatorFn, Validators } from '@angular/forms';
```

**Pull request merged**